### PR TITLE
Reduced span overheads in objective function calculate

### PR DESCRIPTION
**Pull request by ShvetsKS**
it's part of https://github.com/dmlc/xgboost/pull/7192

Before this changes spans were created for each individual sample. Additional overheads of span `operator[]` were also avoided by usage raw pointers.

|       GetGradient     | airline | higgs | 
|------------|---------|---------|
| master, s  | 50.62 | 4.99 | 
| this PR, s | 16.34 | 1.64 | 
| speedup    | 3.1 | 3.04 |

### CODE REVIEW

File: src/objective/regression_obj.cu [hunk 1]
```diff
@@ -73,29 +73,43 @@ class RegLossObj : public ObjFunction {
     additional_input_.HostVector().begin()[1] = scale_pos_weight;
     additional_input_.HostVector().begin()[2] = is_null_weight;
 
-    common::Transform<>::Init([] XGBOOST_DEVICE(size_t _idx,
+    const size_t nthreads = omp_get_max_threads();
+    const size_t n_data_blocks = nthreads;
```
Comment (trivialfis) [inline src/objective/regression_obj.cu:77]: Can we simply use the raw pointer optimization and remove the thread block?  This function is universal to CPU and GPU. (please don't make a separated implementation ;-) )

File: src/objective/regression_obj.cu [hunk 2]
```diff
@@ -73,29 +73,43 @@ class RegLossObj : public ObjFunction {
...
+            bst_float label = labels_ptr[idx];
+            if (label == 1.0f) {
+              w *= _scale_pos_weight;
+            }
+            if (!Loss::CheckLabel(label)) {
+              // If there is an incorrect label, the host code will know.
+              _additional_input[0] = 0;
+            }
+            out_gpair_ptr[idx] = GradientPair(Loss::FirstOrderGradient(p, label) * w,
+                                            Loss::SecondOrderGradient(p, label) * w);
```
Comment (trivialfis) [inline src/objective/regression_obj.cu:106]: Indent.

**Review by trivialfis — CHANGES_REQUESTED** [review]
Thanks for spiting up the PR.

File: src/objective/regression_obj.cu (hunk 1 above)
Comment (trivialfis) [reply src/objective/regression_obj.cu:77]: Or alternatively you can modify the transform function.  But I prefer to keep things simple considering the objective should be a small fraction of overall runtime.

File: src/objective/regression_obj.cu (hunk 2 above)
Comment (ShvetsKS) [reply src/objective/regression_obj.cu]: fixed

File: src/objective/regression_obj.cu (hunk 1 above)
Comment (ShvetsKS) [reply src/objective/regression_obj.cu]: seems no need to modify transform func, added simple support of old behaviour for gpu device (`n_data_blocks == ndata`)

**Comment by ShvetsKS** [comment]
Not sure that https://app.travis-ci.com/github/dmlc/xgboost/jobs/537364519 problem in Travis CI is related to current changes.

**Comment by trivialfis** [comment]
Restarted.

**Comment by trivialfis** [comment]
Hmm ... not sure what's new on Travis.

**Comment by ShvetsKS** [comment]
similar failed test (`tests\python\test_with_sklearn.py::test_estimator_reg`) in master: https://xgboost-ci.net/blue/organizations/jenkins/xgboost-win64/detail/master/948/pipeline, so problem in step `Jenkins Win64: Test` is not related to current changes

**Comment by trivialfis** [comment]
Yeah, handling the CI is the hardest part of maintenance ... it never lacks surprises.

**Comment by napetrov** [comment]
@trivialfis  - should we proceed then with merging this PR?

**Comment by trivialfis** [comment]
Will make some changes before merging.  Thanks for the patience.

**Comment by ShvetsKS** [comment]
> Will make some changes before merging. Thanks for the patience.

@trivialfis seems all checks are passed. Could you take a look at applied comments and help with defining the expected changes?

**Review by trivialfis — COMMENTED** [review]
Sorry for the slow response.  I can merge it as it's but I have to be honest that I'm quite reluctant to do so.  I generated 1e8 samples with a single column for testing the gradient procedure.  The max_depth is set to 1.  I compared the result between using thread block and using a plain parallel loop, the results are 3.21 sec and 7.59 sec.  While the `UpdateOneIter` takes 117 seconds, and again that's with a single feature.  I'm not sure why would you make this specialization on a general elementwise transform function for 4  seconds.

File: src/objective/regression_obj.cu [hunk 3]
```diff
@@ -73,29 +73,46 @@ class RegLossObj : public ObjFunction {
...
+            if (label == 1.0f) {
+              w *= _scale_pos_weight;
+            }
+            if (!Loss::CheckLabel(label)) {
+              // If there is an incorrect label, the host code will know.
+              _additional_input[0] = 0;
+            }
+            out_gpair_ptr[idx] = GradientPair(Loss::FirstOrderGradient(p, label) * w,
+                                              Loss::SecondOrderGradient(p, label) * w);
+            if (out_gpair_ptr[idx].GetHess() < 0.0f) {
```
Comment (trivialfis) [inline src/objective/regression_obj.cu:110]: Could you please share which objective is generating negative gradient?

**Comment by ShvetsKS** [comment]
> Sorry for the slow response. I can merge it as it's but I have to be honest that I'm quite reluctant to do so. I generated 1e8 samples with a single column for testing the gradient procedure. The max_depth is set to 1 and I have 12 threads. I compared the result between using thread block and using a plain parallel loop, the results are 3.21 sec and 7.59 sec. While the `UpdateOneIter` takes 117 seconds, and again that's with a single feature and a single node. I'm not sure why would you make this specialization on a general elementwise transform function for 4 seconds.
> 
> Update: I forgot the gradient index generation, which seems to be taking the majority of time.

For airline dataset for example this PR helps to save 34s (with default params of [benchmark](https://github.com/NVIDIA/gbm-bench )). For current/master version it's not big improvement but with optimized partition, hist building, evaluate and sync kernels `GetGradient` function becomes one of the hotspots.

**Comment by trivialfis** [comment]
Opened a small PR on https://github.com/ShvetsKS/xgboost/pull/3 .

File: src/objective/regression_obj.cu (hunk 3 above)
Comment (ShvetsKS) [reply src/objective/regression_obj.cu:110]: mse, log loss can generate negative gradients (for hessian log-loss e.g.).
But this hessian check was added to avoid similar [here](https://github.com/dmlc/xgboost/blob/master/src/tree/updater_quantile_hist.cc#L483) checks in next PRs (there is no need to read gradient/hessian vector additionally during updater call).

**Comment by ShvetsKS** [comment]
> Opened a small PR on [ShvetsKS#3](https://github.com/ShvetsKS/xgboost/pull/3) .

thanks a lot for corrections!

**Comment by trivialfis** [comment]
No problem, could you please check https://github.com/ShvetsKS/xgboost/pull/3#discussion_r714239072 .

**Review by trivialfis — APPROVED** [review]
